sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -417,6 +417,14 @@ object NullPropagation extends Rule[LogicalPlan] {
         case left :: Literal(null, _) :: Nil => Literal.create(null, e.dataType)
         case _ => e
       }
+
+      // If the value expression is NULL then transform the In expression to
+      // Literal(null)
+      case e @ In(v, list) => v match {
```

Comment:
@cloud-fan Thanks Wenchen for your prompt input as always. Got a question ..

- ConstantFolding rule seems to come after the NullPropagation. Would this still work ?
- The 2nd test case i have added is failing 
   == FAIL: Plans do not match ===
!Filter null IN (1,cast(b#0 as string))   Filter null
  LocalRelation [a#0,b#0,c#0]              LocalRelation [a#0,b#0,c#0]

I am debugging it now. But thought i should check with you first..

Thanks again.